src/m3ninx/index/segment/builder/builder.go

```diff
@@ -90,26 +207,83 @@ func NewBuilderFromDocuments(opts Options) (segment.CloseableDocumentsBuilder, e
 		idSet: NewIDsMap(IDsMapOptions{
 			InitialSize: opts.InitialCapacity(),
 		}),
-		uniqueFields: make([][]uniqueField, 0, concurrency),
-		indexQueues:  make([]chan indexJob, 0, concurrency),
+		shardedFields: &shardedFields{},
 	}
+	// Indiciate we need to spin up workers if we haven't already.
+	globalIndexWorkers.registerBuilder()
+	b.SetIndexConcurrency(opts.Concurrency())
+	return b, nil
+}
 
-	for i := 0; i < concurrency; i++ {
-		indexQueue := make(chan indexJob, indexQueueSize)
-		b.indexQueues = append(b.indexQueues, indexQueue)
-		go b.indexWorker(indexQueue)
+func (b *builder) SetIndexConcurrency(value int) {
+	b.status.Lock()
+	defer b.status.Unlock()
 
+	if b.concurrency == value {
+		return // No-op
+	}
+
+	b.concurrency = value
+
+	// Nothing to migrate, jobs only used during a batch insertion.
+	b.shardedJobs = make([]indexJob, b.concurrency)
+
+	// Take refs to existing fields to migrate.
+	existingUniqueFields := b.shardedFields.uniqueFields
+	existingFields := b.shardedFields.fields
+
+	b.shardedFields.uniqueFields = make([][]uniqueField, 0, b.concurrency)
```

Comment:
nit: should this refresh the existing fields or do we want to explicitly create a new slice here?